### ISSUE TITLE
mypy confuses module with contents

### ISSUE THREAD

**Issue by hauntsaninja** [python/mypy#12681]
Here's a minimal repro (I think every part of this is necessary):

```
.
└── example
    ├── __init__.py
    ├── cell.py
    └── component.py

1 directory, 3 files
```

```
λ cat example/__init__.py
from example.cell import cell
import example.component

λ cat example/cell.py 
def cell() -> str:
    return "asdf"

λ cat example/component.py
import example

x: str = example.cell()
```

And this gives you:
```
λ mypy example
example/component.py:3: error: Module not callable
Found 1 error in 1 file (checked 3 source files)
```

Despite the fact that things work at runtime:
```
λ python -c 'import example; print(example.component.x)'
asdf
```

**Comment by hauntsaninja** [comment]
Note that using `example.cell` when not at module level does the right thing:
```
λ cat example/component.py
import example

def f() -> None:
    x: str = example.cell()

λ mypy example   
Success: no issues found in 3 source files
```

**Comment by bryanforbes** [comment]
I've got another repro:

```
my_thing/
├── __init__.py
└── sub_dir/
    ├── __init__.py
    └── sub_thing.py
tests/
└── test_sub_thing.py
```

**my_thing/sub_dir/sub_thing.py:**
```python
from __future__ import annotations

class SubThing:
    ...

__all__ = ('SubThing',)
```

**my_thing/sub_dir/\_\_init__.py:**
```python
from __future__ import annotations

from .sub_thing import SubThing

def sub_thing() -> SubThing:
    return SubThing()

__all__ = ('sub_thing',)
```

**my_thing/\_\_init__.py:**
```python
from __future__ import annotations

from .sub_dir.sub_thing import SubThing
from .sub_dir import sub_thing

__all__ = ('sub_thing', 'SubThing')
```

**tests/test_sub_thing.py:**
```python
from __future__ import annotations

import my_thing


def test_sub_thing() -> None:
    assert isinstance(my_thing.sub_thing(), my_thing.SubThing)

```

I can reliably reproduce this error by running `mypy my_thing` **and then** running `mypy tests`. If I remove the `mypy` cache files and re-run `mypy tests` first, no issues are found.

**Comment by bryanforbes** [comment]
I forgot to add this earlier: I'm running into this issue working on the types for [pendulum 3.0](https://github.com/sdispater/pendulum/). It would be great to get this fixed for the next release of mypy (or to point me in the right direction of fixing this) to avoid prefixing the similarly named modules with `_`.
